src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi3Driver.Linux.cs

```diff
@@ -358,13 +405,60 @@ private void Initialize()
                     return;
                 }
 
-                int fileDescriptor = Interop.open(GpioMemoryFilePath, FileOpenFlags.O_RDWR | FileOpenFlags.O_SYNC);
+                // try and open /dev/gpiomem
+                fileDescriptor = Interop.open(GpioMemoryFilePath, FileOpenFlags.O_RDWR | FileOpenFlags.O_SYNC);
                 if (fileDescriptor == -1)
                 {
-                    throw new IOException($"Error {Marshal.GetLastWin32Error()} initializing the Gpio driver.");
+                    win32Error = Marshal.GetLastWin32Error();
+
+                    // if the failure is NOT because /dev/gpiomem doesn't exist then throw an exception at this point.
+                    // if it were anything else then it is probably best not to try and use /dev/mem on the basis that
+                    // it would be better to solve the issue rather than use a method that requires root privileges
+                    if (win32Error != Linux_Enoent)
+                    {
+                        throw new IOException($"Error {win32Error} initializing the Gpio driver.");
+                    }
+
+                    // if /dev/gpiomem doesn't seem to be available then let's try /dev/mem
+                    fileDescriptor = Interop.open(MemoryFilePath, FileOpenFlags.O_RDWR | FileOpenFlags.O_SYNC);
+                    if (fileDescriptor == -1)
+                    {
+                        throw new IOException($"Error {Marshal.GetLastWin32Error()} initializing the Gpio driver.");
+                    }
+                    else // success so set the offset into memory of the gpio registers
+                    {
+                        gpioRegisterOffset = ~0U;
+
+                        try
+                        {
+                            // get the periphal base address from the libbcm_host library which is the reccomended way
+                            // according to the RasperryPi website
+                            gpioRegisterOffset = Interop.libbcmhost.bcm_host_get_peripheral_address();
+
+                            // if we get zero back then we use our own internal method. This can happen 
+                            // on a Pi4 if the userland libraries haven't been updated and was fixed in Jul/Aug 2019.
+                            if(gpioRegisterOffset == 0)
+                            {
+                                gpioRegisterOffset = GetPeripheralBaseAddress();
+                            }
+                        }
+                        catch(DllNotFoundException)
```

Comment:
is this exception expected on some PIs? if not I'd rather just let it bubble up

Reply:
On some Pi's that don't use Raspbian then I have found this library missing in the default install. I can't remember of it was the Arm64 Debian or Ubuntu.